core/emitter.js

```diff
@@ -4,14 +4,12 @@ import logger from './logger';
 let debug = logger('quill:events');
 
 const EVENTS = ['selectionchange', 'mousedown', 'mouseup', 'click'];
+const EMITTERS = [];
 
 EVENTS.forEach(function(eventName) {
   document.addEventListener(eventName, (...args) => {
-    [].slice.call(document.querySelectorAll('.ql-container')).forEach((node) => {
```

Comment:
This was added as a memory leak way to add event listeners on global objects. If WeakMap were available it would be used like the comment mentions but since it is not it does this which by the way is how WeakMaps work under the hook.

Reply:
The old method won't work with shadow DOM due to the fact that `querySelectorAll` will not penetrate through shadow roots. 

Though I can see how `EMITTERS` could keep growing if you repeatedly make new instances (and they never be released). Needs some thought..

Reply:
Why couldn't emitters be passed as an initialization argument? Making this global is a recipe for leaks and also makes encapsulating the events for ShadowDOM very hard.  It seems we might be able to make emitter into an interface, then have a `GlobalEventEmitter` and a `ShadownDOMEventEmitter` where Global does the current global things and the ShadowDOMVersion is initialized with the ShadowRoot or editor root.

I'd be down to prototype this, but maybe there is some other consideration.  I'd like to get perspective from yall before working on it.  What do you think?